### PR TITLE
Extract string dependencies into its own files

### PR DESCRIPTION
**Pull request by allredj**
Based on https://github.com/diffblue/cbmc/pull/4692.
Only review last 2 commits.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by allredj** [comment]
rebased onto develop

**Review by allredj — COMMENTED** [review]
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 15f36b6c52d076454a92b27ce6d2dc6ebbda96a4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112897202
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 2fdc6f339517c26461b734ba731360fc2a22da6d).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112900910

File: src/solvers/strings/string_dependencies.h [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+\*******************************************************************/
+
+#ifndef CPROVER_SOLVERS_STRINGS_STRING_DEPENDENCIES_H
+#define CPROVER_SOLVERS_STRINGS_STRING_DEPENDENCIES_H
+
+#include <memory>
+
+#include "string_builtin_function.h"
+#include "string_constraint.h"
+#include "string_constraint_generator.h"
```
Comment (romainbrenguier) [inline src/solvers/strings/string_dependencies.h]: I think this include is not necessary if you forward declare ` string_constraint_generatort`

File: src/solvers/strings/string_dependencies.h [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+\*******************************************************************/
+
+#ifndef CPROVER_SOLVERS_STRINGS_STRING_DEPENDENCIES_H
+#define CPROVER_SOLVERS_STRINGS_STRING_DEPENDENCIES_H
+
+#include <memory>
+
+#include "string_builtin_function.h"
+#include "string_constraint.h"
```
Comment (romainbrenguier) [inline src/solvers/strings/string_dependencies.h]: I don't think this include is used

**Review by romainbrenguier — APPROVED** [review]

File: src/solvers/strings/string_refinement.cpp [hunk 3]
```diff
@@ -31,6 +31,7 @@ Author: Alberto Griggio, alberto.griggio@gmail.com
 
 #include "equation_symbol_mapping.h"
 #include "string_constraint_instantiation.h"
+#include "string_dependencies.h"
```
Comment (antlechner) [inline src/solvers/strings/string_refinement.cpp:34]: This is already included in the header file so it's not needed here.

File: src/solvers/strings/string_dependencies.cpp [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  return builtin_function_nodes.back();
+}
+
+const string_builtin_functiont &string_dependenciest::get_builtin_function(
+  const builtin_function_nodet &node) const
+{
+  return *node.data;
+}
+
+static void for_each_atomic_string(
```
Comment (antlechner) [inline src/solvers/strings/string_dependencies.cpp:100]: :question: Is this function related only to dependencies?

File: src/solvers/strings/string_dependencies.h [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+#include "string_constraint_generator.h"
+
+/// Keep track of dependencies between strings.
+/// Each string points to the builtin_function calls on which it depends.
+/// Each builtin_function points to the strings on which the result depends.
```
Comment (antlechner) [inline src/solvers/strings/string_dependencies.h:21]: :bulb: I think this documentation should answer the question "What do we mean by `dependency` and how do we use dependencies?". An example, or reference to markdown files about the string solver with examples, would be helpful here. I don't know much about the string solver so I'm mostly guessing what any of this might mean. :slightly_smiling_face:

**Review by antlechner — APPROVED** [review]

File: src/solvers/strings/string_dependencies.cpp [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+
+\*******************************************************************/
+
+#include "string_dependencies.h"
+#include <algorithm>
+#include <functional>
+#include <iostream>
+#include <numeric>
+#include <unordered_set>
+#include <util/arith_tools.h>
```
Comment (antlechner) [inline src/solvers/strings/string_dependencies.cpp]: :bulb: I've seen empty lines in other files between STL includes and includes of files from the same project. The coding standard doesn't say anything about this other than the first dependency being the header file for this .cpp file though...
:hammer_and_pick: It looks like the only necessary includes here are
```
#include "string_dependencies.h"
#include <unordered_set>
#include <util/expr_iterator.h>
#include <util/graph.h>
#include <util/make_unique.h>
#include <util/ssa_expr.h>
```

File: src/solvers/strings/string_refinement.h [hunk 7]
```diff
@@ -27,6 +27,7 @@ Author: Alberto Griggio, alberto.griggio@gmail.com
 
 #include "string_constraint.h"
 #include "string_constraint_generator.h"
+#include "string_dependencies.h"
```
Comment (antlechner) [inline src/solvers/strings/string_refinement.h:30]: :pick: Could forward declare `string_dependenciest` instead of including this file.

File: src/solvers/strings/string_dependencies.cpp (hunk 4 above)
Comment (allredj) [reply src/solvers/strings/string_dependencies.cpp:100]: Probably not, but it's not used anywhere else yet.

**Comment by allredj** [comment]
@antlechner @romainbrenguier  I cleaned up the imports.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9e0b64f445849102b4345f6a47b17780b369309e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113098867

File: src/solvers/strings/string_dependencies.h [hunk 8]
```diff
@@ -0,0 +1,198 @@
+/*******************************************************************\
+
+Module: String solver
+
+Author: Diffblue Ltd.
+
+\*******************************************************************/
+
+#ifndef CPROVER_SOLVERS_STRINGS_STRING_DEPENDENCIES_H
```
Comment (peterschrammel) [inline src/solvers/strings/string_dependencies.h:12]: doxygen `\file` is missing.

**Review by peterschrammel — APPROVED** [review]

File: src/solvers/strings/string_dependencies.h (hunk 8 above)
Comment (allredj) [reply src/solvers/strings/string_dependencies.h:12]: @peterschrammel Is that really useful? The class is already documented.

Comment (allredj) [reply src/solvers/strings/string_dependencies.h:12]: I would only repeat the description.

Comment (peterschrammel) [reply src/solvers/strings/string_dependencies.h:12]: @allredj, please check whether it generates an entry next to the file here: http://cprover.diffblue.com/files.html

Comment (allredj) [reply src/solvers/strings/string_dependencies.h:12]: Indeed there's no description on the file view. I added a file documentation for this file.

**Comment by tautschnig** [comment]
@allredj Please rebase now that #4729 is in to make CI happy.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b2bdda568f92983b042fe9b4e8a3548f4555550c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/113725627
